Add test operator in upgrader entry

tugsbayasgalan opened the pull request:
In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69427

Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

tugsbayasgalan committed:
Add test operator in upgrader entry

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

cccclai commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -13,6 +13,10 @@ struct UpgraderEntry {
 };
 
 const static std::unordered_map<std::string, std::vector<UpgraderEntry>> kOperatorVersionMap(
+    {{"aten::_test_serialization_subcmul",
```
It's a bit unintuitive to add a test upgrader in the actual version map. Is the cost ok for us? cc: @iseeyuan

iseeyuan replied:
Is it possible to emplace this test in another header file, and only include it when testing?

tugsbayasgalan replied:
@gmagogsfm  What do you think?

gmagogsfm replied:
It is not good practice to include test-only code in prod code path. One common pattern is to have test-only utilities that allow test cases to register a different set of data (upgraders in this case) so that the tests that cover infrastructure for upgrader support is completely decoupled from the existence of any prod upgraders.

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

gmagogsfm commented on torch/jit/operator_upgraders.py:
```diff
@@ -95,8 +100,9 @@ def collect_available_upgraders():
-            if entry not in available_upgraders_in_version_map:
-                raise AssertionError("The upgrader {} is not registered in the version_map.h")
+            # test entries will be populated in the test cases so we ignore them here
+            if (not entry.startswith("_test")) and (entry not in available_upgraders_in_version_map):
```
Is this needed? `entries` is populated from `globals()`, which shouldn't have any test upgraders in the first place. What am I missing?

tugsbayasgalan replied:
Removed

gmagogsfm commented on test/jit/test_upgraders.py:
```diff
@@ -37,6 +37,23 @@ def f():
         self.assertTrue(upgraders_size == upgraders_size_second_time)
         self.assertTrue(upgraders_dump == upgraders_dump_second_time)
 
+    def test_add_value_to_version_map(self):
```
This method seems to be testing the test-only methods themselves. Could we use the `_add_test_entry` method to actually test other part of upgrader processing code? Like registering an upgrader and apply it on a piece of old IR, which gets upgraded etc.

gmagogsfm replied:
Ping

tugsbayasgalan replied:
These PRs come before the PRs where we actually replace the old IR. (https://github.com/pytorch/pytorch/pull/68358)
In there, I will be using these utility methods to test our change.

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).

gmagogsfm commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -12,7 +12,7 @@ struct UpgraderEntry {
 };
 
-const static std::unordered_map<std::string, std::vector<UpgraderEntry>> kOperatorVersionMap(
+static std::unordered_map<std::string, std::vector<UpgraderEntry>> kOperatorVersionMap(
```
I think `k` prefix is only for constants.

tugsbayasgalan replied:
sounds good. will do

tugsbayasgalan replied:
done

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -47,5 +47,13 @@ get_operator_version_map() {
+  kOperatorVersionMap[op_name].push_back(entry);
+}
+
+void remove_test_entry(std::string op_name) {
```
done

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -47,5 +47,13 @@ get_operator_version_map() {
   return kOperatorVersionMap;
 }
 
+void add_test_entry(std::string op_name, UpgraderEntry entry) {
```
done

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

Commit:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add test operator in upgrader entry"

In this PR, we add utility methods to remove/add test operator entries to the operator version map we track. 


Differential Revision: [D32867984](https://our.internmc.facebook.com/intern/diff/D32867984)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32867984).
